File: platform/daemon/src/pipeline/dreaming-worker.ts
/**
 * Dreaming worker — periodically checks token threshold and triggers
 * consolidation passes. Manages the dreaming lifecycle as a daemon
 * background task.
 */

import type { DreamingConfig } from "@signet/core";
import type { DbAccessor } from "../db-accessor";
import { getQueueHealth } from "../diagnostics";
import { getOrCreateInferenceRouter } from "../inference-router";
import { logger } from "../logger";
import { isSystemPressureHigh } from "../system-pressure";
import {
	type DreamingAgentExecutor,
	type DreamingMode,
	type DreamingPassFocus,
	createDreamingPass,
	dreamingFocusOfMode,
	enqueueDreamingHygieneAttention,
	getDreamingEpisodicTokenBacklog,
	isDreamingHaltActive,
	recordDreamingFailure,
	runDreamingAgentPass,
	selectDreamingPassMode,
	shouldTriggerDreaming,
} from "./dreaming";
import { getDreamingAttentionInDb } from "./dreaming-attention";

/** Thrown when a trigger is attempted while a pass is already in-flight. */
export class AlreadyRunningError extends Error {
	constructor() {
		super("A dreaming pass is already running");
		this.name = "AlreadyRunningError";
	}
}

export interface DreamingWorkerHandle {
	stop(): void;
	/** Force-trigger a pass synchronously (CLI / testing). */
	trigger(
		mode: DreamingMode,
		agentId?: string,
	): Promise<{ passId: string; applied: number; skipped: number; failed: number; summary: string }>;
	/**
	 * Fire-and-forget trigger: creates the pass record synchronously
	 * (so the passId is returned immediately), then runs the pass in the
	 * background. Callers should poll GET /api/dream/status for completion.
	 * Throws AlreadyRunningError if a pass is already active.
	 */
	triggerAsync(mode: DreamingMode, agentId?: string): string;
	readonly running: boolean;
	readonly activeAgentId: string | null;
	/**
	 * Resolves when the in-flight pass completes (or is null when idle).
	 * Await this (with a timeout) during shutdown before closing the DB.
	 */
	readonly activePass: Promise<unknown> | null;
}

export interface DreamingWorkerOptions {
	/** Test seam; production always uses the configured inference router. */
	readonly executorFactory?: (agentId: string) => DreamingAgentExecutor;
	/** Scoped connection details for ACPX's temporary MCP server. */
	readonly acpxMcp?: {
		readonly daemonUrl: string;
		readonly authorizationTokenForAgent?: (agentId: string) => string | undefined;
	};
}

const CHECK_INTERVAL_MS = 5 * 60 * 1000; // 5 min
const AGENT_SCOPE_SNAPSHOT_REFRESH_MS = 30 * 60 * 1000; // 30 min

/** Dreaming is deferrable work. Yield an entire sweep while live queues are under pressure. */
export function shouldDeferDreamingSweep(accessor: DbAccessor): boolean {
	return accessor.withReadDb((db) => getQueueHealth(db).status !== "healthy");
}

function normalizeAgentId(agentId: string | undefined, fallback: string): string {
	const trimmed = agentId?.trim();
	return trimmed ? trimmed : fallback;
}

export function getDreamingWorkerAgentIds(accessor: DbAccessor, defaultAgentId: string): readonly string[] {
	return accessor.withReadDb((db) => {
		// UNION ALL + app-side dedup instead of UNION: the caller collapses
		// rows into a Set, so the cross-branch sort/merge UNION performs is
		// pure waste. UNION ALL concatenates the per-table index scans
		// (every branch resolves through an agent_id-prefix or covering
		// index), bounding the query by index size rather than table
		// content (#1094).
		const rows = db
			.prepare(
				`SELECT id AS id FROM agents
				 UNION ALL
				 SELECT DISTINCT agent_id AS id FROM dreaming_state
				 UNION ALL
				 SELECT DISTINCT agent_id AS id FROM dreaming_passes
				 UNION ALL
				 SELECT DISTINCT agent_id AS id FROM memories WHERE is_deleted = 0
				 UNION ALL
				 SELECT DISTINCT agent_id AS id FROM session_summaries
				 UNION ALL
				 SELECT DISTINCT agent_id AS id FROM memory_artifacts WHERE is_deleted = 0
				 UNION ALL
				 SELECT DISTINCT agent_id AS id FROM session_transcripts
				 UNION ALL
				 SELECT DISTINCT agent_id AS id FROM dreaming_attention WHERE resolved_at IS NULL
				 UNION ALL
				 SELECT DISTINCT agent_id AS id FROM entities`,
			)
			.all() as Array<{ id: string | null }>;
		const ids = new Set<string>([defaultAgentId]);
		for (const row of rows) {
			const id = normalizeAgentId(row.id ?? undefined, "");
			if (id) ids.add(id);
		}
		return [...ids].sort();
	});
}

/**
 * Bounded agent-scope discovery: resolves the union query on a refresh
 * cadence and serves the snapshot between refreshes. Periodic dreaming is
 * deferrable, so a stale list only delays work for a brand-new scope by one
 * refresh window instead of re-running the 9-table union on every sweep
 * (#1059).
 */
export function createAgentScopeSnapshot(
	refreshMs: number,
	resolve: () => readonly string[],
	now: () => number = Date.now,
): () => readonly string[] {
	let snapshot: readonly string[] | null = null;
	let at = 0;
	return () => {
		const t = now();
		if (snapshot === null || t - at >= refreshMs) {
			snapshot = resolve();
			at = t;
		}
		return snapshot;
	};
}

/**
 * The runbook for the next scheduled sweep pass (#1098): read the pending
 * work across every scope, then pick hygiene/content — alternating when both
 * kinds are pending so content gets a guaranteed turn even while the hygiene
 * queue stays full. Shared between check() and tests.
 */
export function selectDreamingCheckMode(
	accessor: DbAccessor,
	scopes: readonly string[],
	lastScheduled: DreamingPassFocus | null,
): DreamingMode {
	const hasPendingAttention = scopes.some((scope) =>
		accessor.withReadDb((db) => getDreamingAttentionInDb(db, scope, 1).length > 0),
	);
	const hasBacklog = scopes.some((scope) => getDreamingEpisodicTokenBacklog(accessor, scope) > 0);
	return selectDreamingPassMode(lastScheduled, hasPendingAttention, hasBacklog);
}

export function startDreamingWorker(
	accessor: DbAccessor,
	cfg: DreamingConfig,
	agentsDir: string,
	defaultAgentId: string,
	options: DreamingWorkerOptions = {},
): DreamingWorkerHandle {
	let timer: ReturnType<typeof setTimeout> | null = null;
	let active = false;
	let activeAgent: string | null = null;
	let stopped = false;
	let activePassPromise: Promise<unknown> | null = null;
	// The last focused runbook the periodic sweep scheduled, used to
	// alternate hygiene → content → hygiene → … when both kinds of work are
	// pending (#1098). Explicit triggers do not touch it.
	let nextScheduledFocus: DreamingPassFocus | null = null;
	const getAgentScopes = createAgentScopeSnapshot(AGENT_SCOPE_SNAPSHOT_REFRESH_MS, () =>
		getDreamingWorkerAgentIds(accessor, defaultAgentId),
	);
	const executorForAgent = (agentId: string): DreamingAgentExecutor => {
		const factory = options.executorFactory;
		if (factory) return factory(agentId);
		// getOrCreateInferenceRouter is a singleton accessor; resolving here
		// (instead of a nullable eager handle) keeps the closure free of
		// non-null assertions.
		const router = getOrCreateInferenceRouter(agentsDir);
		return {
			async run(input) {
				const result = await router.runAgent(
					{
						agentId,
						operation: "memory_extraction",
						promptPreview: input.prompt.slice(0, 8000),
					},
					input.prompt,
					input.tools,
					{
						timeoutMs: input.timeoutMs,
						maxTokens: input.maxTokens,
						...(options.acpxMcp
							? {
									acpxMcp: {
										agentId,
										passId: input.passId,
										daemonUrl: options.acpxMcp.daemonUrl,
										authorizationToken: options.acpxMcp.authorizationTokenForAgent?.(agentId),
									},
								}
							: {}),
					},
				);
				if (!result.ok) {
					const attempts = Array.isArray(result.error.details?.attempts)
						? result.error.details.attempts
								.map((attempt) => {
									if (!attempt || typeof attempt !== "object") return "unknown target";
									const value = attempt as { targetRef?: unknown; error?: unknown };
									return `${typeof value.targetRef === "string" ? value.targetRef : "unknown"}: ${typeof value.error === "string" ? value.error : "failed"}`;
								})
								.join("; ")
						: "";
					throw new Error(attempts ? `${result.error.message} (${attempts})` : result.error.message);
				}
				return { summary: `Dreaming agent completed through ${result.value.decision.targetRef}` };
			},
		};
	};

	// Sweep orphaned passes from unclean shutdown: any 'running' record
	// was left by a crash or forced stop — mark it failed
	// so the status API doesn't show a forever-running ghost pass.
	accessor.withWriteTx((db) => {
		const orphaned = db
			.prepare(
				`UPDATE dreaming_passes
				 SET status = 'failed',
				     completed_at = datetime('now'),
				     error = 'Orphaned by daemon restart'
				 WHERE status = 'running'`,
			)
			.run();
		if (orphaned.changes > 0) {
			logger.warn("dreaming-worker", `Swept ${orphaned.changes} orphaned running pass(es) from prior shutdown`);
		}
	});

	async function runPass(
		runAgentId: string,
		mode: DreamingMode,
		existingPassId?: string,
		scopes?: readonly string[],
	): Promise<{ passId: string; applied: number; skipped: number; failed: number; summary: string }> {
		if (active) throw new AlreadyRunningError();
		active = true;
		activeAgent = runAgentId;
		// Periodic sweeps pass their snapshot through; explicit triggers and
		// CLI passes resolve fresh so operator intent is never stale.
		const passScopes = scopes ?? getDreamingWorkerAgentIds(accessor, defaultAgentId);
		const p = runDreamingAgentPass(
			accessor,
			executorForAgent(runAgentId),
			cfg,
			agentsDir,
			runAgentId,
			passScopes,
			mode,
			existingPassId,
		);
		activePassPromise = p;
		try {
			return await p;
		} catch (e) {
			recordDreamingFailure(accessor, runAgentId);
			throw e;
		} finally {
			active = false;
			activeAgent = null;
			activePassPromise = null;
		}
	}

	async function check(): Promise<void> {
		if (stopped || active) return;
		if (isSystemPressureHigh()) return;
		if (shouldDeferDreamingSweep(accessor)) {
			logger.info("dreaming-worker", "Deferring dreaming sweep while queues are under pressure");
			return;
		}

		// One Dreaming universe: a single pass covers every agent scope. The
		// sweep runs one pass when any scope has attention or a backlog; the
		// pass itself addresses scopes via the per-call agentId on its tools.
		const scopes = getAgentScopes();
		let triggered = false;
		for (const scopeId of scopes) {
			if (stopped || active) return;
			// A halted scope must not burn the 6-query hygiene scan and the
			// episodic backlog read on every sweep: skip straight past it.
			if (isDreamingHaltActive(accessor, scopeId)) continue;
			try {
				enqueueDreamingHygieneAttention(accessor, scopeId);
				const episodicTokens = getDreamingEpisodicTokenBacklog(accessor, scopeId);
				if (!shouldTriggerDreaming(accessor, cfg, scopeId, Date.now(), episodicTokens)) continue;
				triggered = true;
				logger.info("dreaming-worker", "Episodic evidence threshold reached, starting dreaming pass", {
					scopeId,
					episodicTokens,
					threshold: cfg.tokenThreshold,
				});
				break;
			} catch (e) {
				if (e instanceof AlreadyRunningError) return;
				logger.error("dreaming-worker", "Dreaming scope check failed", undefined, {
					agentId: scopeId,
					error: e instanceof Error ? e.message : String(e),
				});
			}
		}
		if (!triggered) return;
		// #1098: with the hygiene queue perpetually full, every pass used to
		// drain flags first and run out of budget before content work. Give
		// content a guaranteed turn: alternate the runbook per check cycle
		// when both kinds of work are pending; run the only-pending kind
		// directly otherwise.
		const mode = selectDreamingCheckMode(accessor, scopes, nextScheduledFocus);
		nextScheduledFocus = dreamingFocusOfMode(mode) ?? nextScheduledFocus;
		await runPass(defaultAgentId, mode, undefined, scopes);
	}

	function schedule(): void {
		if (stopped) return;
		timer = setTimeout(async () => {
			await check();
			schedule();
		}, CHECK_INTERVAL_MS);
	}

	// Start the periodic check. Hygiene attention is enqueued during regular
	// check() ticks, NOT here — the hygiene scan does 6 SQL queries per agent
	// that can take minutes on large graphs (30k entities = 2s/query on cold
	// cache). Running it at startup blocks the event loop before the HTTP
	// server binds, making the daemon appear to fail to start.
	schedule();

	// Warm the agent-scope snapshot shortly after startup instead of letting
	// the first check (5 min out) resolve it cold on the main loop
	// (#1094). The union itself is index-fast; the warm-up moves first
	// resolution off the sweep, surfaces DB problems before the first
	// check, and the delay lands after the HTTP server binds.
	let warmupTimer: ReturnType<typeof setTimeout> | null = setTimeout(() => {
		warmupTimer = null;
		if (stopped) return;
		try {
			const scopes = getAgentScopes();
			logger.info("dreaming-worker", "Agent scope snapshot primed", { scopes: scopes.length });
		} catch (e) {
			logger.warn("dreaming-worker", "Agent scope snapshot warm-up failed; first check will retry", {
				error: e instanceof Error ? e.message : String(e),
			});
		}
	}, 15_000);

	logger.info("dreaming-worker", "Dreaming worker started", {
		threshold: cfg.tokenThreshold,
	});

	return {
		// Cancels the timer but does NOT await an in-flight pass.
		// An active pass will complete (or fail) asynchronously; the
		// `stopped` flag prevents new passes from being scheduled.
		stop() {
			stopped = true;
			if (timer) {
				clearTimeout(timer);
				timer = null;
			}
			if (warmupTimer) {
				clearTimeout(warmupTimer);
				warmupTimer = null;
			}
		},

		trigger(mode: DreamingMode, agentId?: string) {
			return runPass(normalizeAgentId(agentId, defaultAgentId), mode);
		},

		triggerAsync(mode: DreamingMode, agentId?: string): string {
			if (active) throw new AlreadyRunningError();
			const runAgentId = normalizeAgentId(agentId, defaultAgentId);
			const passId = createDreamingPass(accessor, runAgentId, mode);
			active = true;
			activeAgent = runAgentId;
			const p = runDreamingAgentPass(
				accessor,
				executorForAgent(runAgentId),
				cfg,
				agentsDir,
				runAgentId,
				getDreamingWorkerAgentIds(accessor, defaultAgentId),
				mode,
				passId,
			);
			activePassPromise = p;
			p.catch((e) => {
				recordDreamingFailure(accessor, runAgentId);
				logger.error("dreaming-worker", "Async trigger failed", undefined, {
					agentId: runAgentId,
					passId,
					error: e instanceof Error ? e.message : String(e),
				});
			}).finally(() => {
				active = false;
				activeAgent = null;
				activePassPromise = null;
			});
			return passId;
		},

		get running() {
			return active;
		},

		get activeAgentId() {
			return activeAgent;
		},

		get activePass() {
			return activePassPromise;
		},
	};
}

File: platform/daemon-rs/crates/signet-pipeline/src/dreaming_worker.rs
//! Dreaming worker runtime parity for the TypeScript daemon.
//!
//! This module owns the background scheduler and manual trigger semantics around
//! the lower-level dreaming pass implementation in [`crate::dreaming`]. It
//! mirrors `platform/daemon/src/pipeline/dreaming-worker.ts`: discover all
//! data-bearing agents, sweep orphaned running passes on start, serialize pass
//! execution, and support fire-and-forget manual triggers that return the pass
//! id after the DB row is created.

use std::path::PathBuf;
use std::sync::{Arc, Mutex};
use std::time::Duration;

use rusqlite::params;
use signet_core::db::{DbPool, Priority};
use thiserror::Error;
use tokio::task::JoinHandle;
use tracing::{error, info, warn};

use crate::dreaming::{
    DreamingConfig, DreamingMode, DreamingPassResult, LlmGenerateFn, create_dreaming_pass,
    get_dreaming_state, record_dreaming_failure, run_dreaming_pass, should_trigger_dreaming,
};

const CHECK_INTERVAL_MS: u64 = 5 * 60 * 1_000;

#[derive(Debug, Error)]
pub enum DreamingWorkerError {
    #[error("A dreaming pass is already running")]
    AlreadyRunning,
    #[error("dreaming pass failed: {0}")]
    PassFailed(String),
    #[error("database error: {0}")]
    Database(String),
}

#[derive(Debug, Clone)]
pub struct DreamingWorkerConfig {
    pub check_interval: Duration,
}

impl Default for DreamingWorkerConfig {
    fn default() -> Self {
        Self {
            check_interval: Duration::from_millis(CHECK_INTERVAL_MS),
        }
    }
}

#[derive(Debug, Default)]
struct WorkerState {
    active: bool,
    active_agent_id: Option<String>,
    active_pass_id: Option<String>,
    stopped: bool,
}

/// Handle for a running dreaming worker.
pub struct DreamingWorkerHandle {
    pool: DbPool,
    generator: Arc<dyn LlmGenerateFn + Send + Sync>,
    dreaming_config: DreamingConfig,
    agents_dir: String,
    default_agent_id: String,
    state: Arc<Mutex<WorkerState>>,
    scheduler: Mutex<Option<JoinHandle<()>>>,
}

fn normalize_agent_id(agent_id: Option<&str>, fallback: &str) -> String {
    let trimmed = agent_id.unwrap_or_default().trim();
    if trimmed.is_empty() {
        fallback.to_string()
    } else {
        trimmed.to_string()
    }
}

/// Discover agents the periodic dreaming worker should consider.
///
/// This is a direct Rust port of `getDreamingWorkerAgentIds`: registered
/// agents plus any agent with dreaming state/passes, memories, summaries, or
/// graph entities, with the configured default agent always included.
pub async fn get_dreaming_worker_agent_ids(pool: &DbPool, default_agent_id: &str) -> Vec<String> {
    let fallback = default_agent_id.to_string();
    pool.read(move |conn| {
        let mut stmt = conn.prepare(
            "SELECT id FROM agents
             UNION
             SELECT DISTINCT agent_id AS id FROM dreaming_state
             UNION
             SELECT DISTINCT agent_id AS id FROM dreaming_passes
             UNION
             SELECT DISTINCT agent_id AS id FROM memories WHERE is_deleted = 0
             UNION
             SELECT DISTINCT agent_id AS id FROM session_summaries
             UNION
             SELECT DISTINCT agent_id AS id FROM entities",
        )?;
        let rows = stmt
            .query_map([], |row| row.get::<_, Option<String>>(0))?
            .collect::<rusqlite::Result<Vec<_>>>()?;
        let mut ids = rows
            .into_iter()
            .filter_map(|id| {
                let normalized = normalize_agent_id(id.as_deref(), "");
                (!normalized.is_empty()).then_some(normalized)
            })
            .collect::<Vec<_>>();
        ids.push(fallback);
        ids.sort();
        ids.dedup();
        Ok(ids)
    })
    .await
    .unwrap_or_else(|e| {
        warn!(error = %e, "failed to discover dreaming worker agents");
        vec![default_agent_id.to_string()]
    })
}

async fn sweep_orphaned_passes(pool: &DbPool) {
    match pool
        .write(Priority::Low, |conn| {
            let changes = conn.execute(
                "UPDATE dreaming_passes
                 SET status = 'failed',
                     completed_at = datetime('now'),
                     error = 'Orphaned by daemon restart'
                 WHERE status = 'running'",
                [],
            )?;
            Ok(serde_json::json!({ "changes": changes }))
        })
        .await
    {
        Ok(value) => {
            let changes = value["changes"].as_u64().unwrap_or(0);
            if changes > 0 {
                warn!(
                    changes,
                    "swept orphaned running dreaming pass(es) from prior shutdown"
                );
            }
        }
        Err(e) => warn!(error = %e, "failed to sweep orphaned dreaming passes"),
    }
}

async fn mark_pass_failed(pool: &DbPool, pass_id: &str, error_message: &str) {
    let pass_id = pass_id.to_string();
    let error_message = error_message.to_string();
    if let Err(e) = pool
        .write(Priority::Low, move |conn| {
            conn.execute(
                "UPDATE dreaming_passes
                 SET status = 'failed', completed_at = datetime('now'), error = ?1
                 WHERE id = ?2 AND status = 'running'",
                params![error_message, pass_id],
            )?;
            Ok(serde_json::Value::Null)
        })
        .await
    {
        warn!(error = %e, "failed to mark dreaming pass failed");
    }
}

impl DreamingWorkerHandle {
    pub fn running(&self) -> bool {
        self.state.lock().expect("dreaming worker state").active
    }

    pub fn active_agent_id(&self) -> Option<String> {
        self.state
            .lock()
            .expect("dreaming worker state")
            .active_agent_id
            .clone()
    }

    pub fn active_pass_id(&self) -> Option<String> {
        self.state
            .lock()
            .expect("dreaming worker state")
            .active_pass_id
            .clone()
    }

    /// Cancel future scheduling. An in-flight pass continues asynchronously,
    /// matching the TypeScript worker shutdown contract.
    pub fn stop(&self) {
        {
            let mut state = self.state.lock().expect("dreaming worker state");
            state.stopped = true;
        }
        if let Some(handle) = self.scheduler.lock().expect("dreaming scheduler").take() {
            handle.abort();
        }
    }

    /// Force-trigger a pass and await completion.
    pub async fn trigger(
        &self,
        mode: DreamingMode,
        agent_id: Option<&str>,
    ) -> Result<DreamingPassResult, DreamingWorkerError> {
        let run_agent_id = normalize_agent_id(agent_id, &self.default_agent_id);
        self.run_pass(run_agent_id, mode, None).await
    }

    /// Fire-and-forget trigger. The pass row is created before this method
    /// returns, so callers can poll status by the returned pass id.
    pub async fn trigger_async(
        &self,
        mode: DreamingMode,
        agent_id: Option<&str>,
    ) -> Result<String, DreamingWorkerError> {
        let run_agent_id = normalize_agent_id(agent_id, &self.default_agent_id);
        {
            let mut state = self.state.lock().expect("dreaming worker state");
            if state.active {
                return Err(DreamingWorkerError::AlreadyRunning);
            }
            state.active = true;
            state.active_agent_id = Some(run_agent_id.clone());
            state.active_pass_id = None;
        }

        let pass_id = create_dreaming_pass(&self.pool, &run_agent_id, &mode).await;
        {
            let mut state = self.state.lock().expect("dreaming worker state");
            state.active_pass_id = Some(pass_id.clone());
        }

        let pool = self.pool.clone();
        let generator = self.generator.clone();
        let cfg = self.dreaming_config.clone();
        let agents_dir = self.agents_dir.clone();
        let state = self.state.clone();
        let pass_id_for_task = pass_id.clone();
        tokio::spawn(async move {
            let result = run_dreaming_pass(
                &pool,
                generator.as_ref(),
                &cfg,
                &agents_dir,
                &run_agent_id,
                &mode,
                Some(&pass_id_for_task),
            )
            .await;
            if let Err(e) = result {
                record_dreaming_failure(&pool, &run_agent_id).await;
                mark_pass_failed(&pool, &pass_id_for_task, &e).await;
                error!(agent_id = %run_agent_id, pass_id = %pass_id_for_task, error = %e, "async dreaming trigger failed");
            }
            let mut guard = state.lock().expect("dreaming worker state");
            guard.active = false;
            guard.active_agent_id = None;
            guard.active_pass_id = None;
        });

        Ok(pass_id)
    }

    async fn run_pass(
        &self,
        run_agent_id: String,
        mode: DreamingMode,
        existing_pass_id: Option<String>,
    ) -> Result<DreamingPassResult, DreamingWorkerError> {
        {
            let mut state = self.state.lock().expect("dreaming worker state");
            if state.active {
                return Err(DreamingWorkerError::AlreadyRunning);
            }
            state.active = true;
            state.active_agent_id = Some(run_agent_id.clone());
            state.active_pass_id = existing_pass_id.clone();
        }

        let result = run_dreaming_pass(
            &self.pool,
            self.generator.as_ref(),
            &self.dreaming_config,
            &self.agents_dir,
            &run_agent_id,
            &mode,
            existing_pass_id.as_deref(),
        )
        .await;

        let mut pass_id_to_fail = existing_pass_id;
        if let Ok(pass) = &result {
            pass_id_to_fail = Some(pass.pass_id.clone());
        }

        if let Err(e) = &result {
            record_dreaming_failure(&self.pool, &run_agent_id).await;
            if let Some(pass_id) = pass_id_to_fail.as_deref() {
                mark_pass_failed(&self.pool, pass_id, e).await;
            }
        }

        {
            let mut state = self.state.lock().expect("dreaming worker state");
            state.active = false;
            state.active_agent_id = None;
            state.active_pass_id = None;
        }

        result.map_err(DreamingWorkerError::PassFailed)
    }
}

async fn check_once(
    pool: DbPool,
    generator: Arc<dyn LlmGenerateFn + Send + Sync>,
    dreaming_config: DreamingConfig,
    agents_dir: String,
    default_agent_id: String,
    state: Arc<Mutex<WorkerState>>,
) {
    if state.lock().expect("dreaming worker state").active {
        return;
    }

    for run_agent_id in get_dreaming_worker_agent_ids(&pool, &default_agent_id).await {
        {
            let guard = state.lock().expect("dreaming worker state");
            if guard.stopped || guard.active {
                return;
            }
        }

        let state_snapshot = get_dreaming_state(&pool, &run_agent_id).await;
        let is_first =
            state_snapshot.last_pass_at.is_none() && dreaming_config.backfill_on_first_run;
        let mode = if is_first {
            DreamingMode::Compact
        } else {
            DreamingMode::Incremental
        };

        if !should_trigger_dreaming(&pool, &dreaming_config, &run_agent_id).await {
            continue;
        }

        info!(
            agent_id = %run_agent_id,
            tokens = state_snapshot.tokens_since_last_pass,
            threshold = dreaming_config.token_threshold,
            mode = ?mode,
            "token threshold reached, starting dreaming pass"
        );

        let handle = DreamingWorkerHandle {
            pool: pool.clone(),
            generator: generator.clone(),
            dreaming_config: dreaming_config.clone(),
            agents_dir: agents_dir.clone(),
            default_agent_id: default_agent_id.clone(),
            state: state.clone(),
            scheduler: Mutex::new(None),
        };
        match handle.run_pass(run_agent_id.clone(), mode, None).await {
            Ok(_) => {}
            Err(DreamingWorkerError::AlreadyRunning) => return,
            Err(e) => error!(agent_id = %run_agent_id, error = %e, "dreaming check failed"),
        }
    }
}

/// Start the dreaming worker scheduler.
pub async fn start_dreaming_worker(
    pool: DbPool,
    generator: Arc<dyn LlmGenerateFn + Send + Sync>,
    dreaming_config: DreamingConfig,
    agents_dir: impl Into<PathBuf>,
    default_agent_id: impl Into<String>,
    worker_config: DreamingWorkerConfig,
) -> DreamingWorkerHandle {
    sweep_orphaned_passes(&pool).await;

    let agents_dir = agents_dir.into().display().to_string();
    let default_agent_id = default_agent_id.into();
    let state = Arc::new(Mutex::new(WorkerState::default()));

    let scheduler_pool = pool.clone();
    let scheduler_generator = generator.clone();
    let scheduler_cfg = dreaming_config.clone();
    let scheduler_agents_dir = agents_dir.clone();
    let scheduler_default_agent = default_agent_id.clone();
    let scheduler_state = state.clone();
    let check_interval = worker_config.check_interval;
    let scheduler = tokio::spawn(async move {
        loop {
            tokio::time::sleep(check_interval).await;
            if scheduler_state
                .lock()
                .expect("dreaming worker state")
                .stopped
            {
                break;
            }
            check_once(
                scheduler_pool.clone(),
                scheduler_generator.clone(),
                scheduler_cfg.clone(),
                scheduler_agents_dir.clone(),
                scheduler_default_agent.clone(),
                scheduler_state.clone(),
            )
            .await;
        }
    });

    info!(
        threshold = dreaming_config.token_threshold,
        "dreaming worker started"
    );

    DreamingWorkerHandle {
        pool,
        generator,
        dreaming_config,
        agents_dir,
        default_agent_id,
        state,
        scheduler: Mutex::new(Some(scheduler)),
    }
}
